pkg/providers/vault/client.go

```diff
@@ -27,32 +29,136 @@ import (
 	"github.com/edgexfoundry/go-mod-secrets/pkg"
 )
 
+const (
+	AuthTypeHeader = "X-Vault-Token"
+)
+
 // Client defines the behavior for interacting with the Vault REST secret key/value store via HTTP(S).
 type Client struct {
 	HttpConfig SecretConfig
 	HttpCaller Caller
+
+	// internal member variables
+	lc loggingClient
 }
 
-// NewSecretClient constructs a SecretClient which communicates with Vault via HTTP(S)
-func NewSecretClient(config SecretConfig) (pkg.SecretClient, error) {
-	httpClient, err := createHTTPClient(config)
+func (c Client) refreshToken(ctx context.Context, errChan chan<- error) error {
+	token := c.HttpConfig.Authentication.AuthToken
+
+	tokenData, err := c.getTokenLookupResponseData()
+
 	if err != nil {
-		return Client{}, err
+		return err
 	}
 
-	if config.RetryWaitPeriod != "" {
-		retryTimeDuration, err := time.ParseDuration(config.RetryWaitPeriod)
-		if err != nil {
-			return nil, err
+	if !tokenData.Data.Renewable {
+		// token is not renewable, log warning and return
+		c.lc.Warn(fmt.Sprintf("token '%s' is not renewable from the secret store", token))
+		return nil
+	}
+
+	// the renew interval is half of period value
+	tokenPeriod := time.Duration(tokenData.Data.Period) * time.Second
+	renewInterval := tokenPeriod / 2
+	if renewInterval <= 0 {
+		// no renew
+		c.lc.Warn(fmt.Sprintf("token '%s' renewInterval is 0, no token renewal", token))
+		return nil
+	}
+
+	ttl := time.Duration(tokenData.Data.Ttl) * time.Second
+
+	// if the current time-to-live is already less than the half of period
+	// need to renew the token right away
+	if ttl <= renewInterval {
+		// call renew self api
+		c.lc.Info("ttl already <= half of the renewal period")
+		if err := c.renewToken(); err != nil {
+			return err
 		}
-		config.retryWaitPeriodTime = retryTimeDuration
 	}
 
-	return Client{
-		HttpConfig: config,
-		HttpCaller: httpClient,
-	}, nil
+	// goroutine to periodically renew the service token based on renewInterval
+	go func() {
+		ticker := time.NewTicker(renewInterval)
+		for {
+			select {
+			case <-ctx.Done():
+				ticker.Stop()
+				c.lc.Info("context cancelled, dismiss the token renewal process")
+				return
```

Comment:
Who is responsible from removing the cancel function from `tokenCancelFunc` if indeed the cancel function is called?

Reply:
golang's garbage collector

Reply:
Is there a test for it? (looking up the cancel function can calling it twice?)

Reply:
isn't the unit test TestMultipleTokenRenewals() with test name "New secret client with the same first token again" doing that test?  The first call will create that cancel func and the second call will cancel that.  You probably can implement a version of logger or just use EdgeX's loggingClient with StdOut to observe the calling behavior.

Reply:
Here is the scenario I was thinking:
- Create a secret client
- Have refreshToken() fail
- Create a second secret client with the same token
- Second secret client cancel()'s the first one again before calling refreshToken()

Reply:
I think the current implementation of NewSecretClient() already is doing that scenario and cancels  the first one b4 calling refreshToken() call see line 82-88 of factory.go

Reply:
run this testing code:
```go
func TestMultipleClientsTokenRenewals(t *testing.T) {
	// setup
	tokenPeriod := 6
	var tokenDataMap sync.Map

	// expired token
	tokenDataMap.Store("expiredToken", TokenLookupMetadata{
		Renewable: true,
		Ttl:       0,
		Period:    tokenPeriod,
	})

	server := getMockTokenServer(&tokenDataMap)
	defer server.Close()

	serverUrl, err := url.Parse(server.URL)
	if err != nil {
		t.Errorf("error on parsing server url %s: %s", server.URL, err)
	}
	host, port, _ := net.SplitHostPort(serverUrl.Host)
	portNum, _ := strconv.Atoi(port)

	bkgCtx := context.Background()

	// error channel to receive any error from background process
	errCh := make(chan error, 1)
	go func() {
		tokenErr := <-errCh

		if tokenErr != nil {
			t.Errorf("error on handling token renewal: %s", tokenErr)
		}
	}()

	mockLogger := NewMockClient()
	factory := NewSecretClientFactory()
	cfgHTTP := SecretConfig{
		Host:           host,
		Port:           portNum,
		Protocol:       "http",
		Authentication: AuthenticationInfo{AuthToken: "expiredToken"},
	}

	_, err = factory.NewSecretClient(bkgCtx, cfgHTTP, mockLogger, errCh)
	// it will fail since the token is expired
	if err == nil {
		t.Errorf("expecting an error for expired token")
	}

	// create a second secret client with the same expired token
	_, err = factory.NewSecretClient(bkgCtx, cfgHTTP, mockLogger, errCh)
	if err == nil {
		t.Errorf("expecting an error for expired token")
	} else {
		fmt.Println(err)
	}

	time.Sleep(5 * time.Second)
}
```
and here is the console output I saw:
```txt
=== RUN   TestMultipleClientsTokenRenewals
HTTP response with status code 403, message: failed to renew token
--- PASS: TestMultipleClientsTokenRenewals (5.00s)
PASS
ok  	github.com/edgexfoundry/go-mod-secrets/pkg/providers/vault	5.005s
Success: Tests passed.
```

Reply:
Ok.  Will take you word for it.  Just wanted to make sure that case was covered.

Reply:
thanks for details and being very caution-awared! Appreciate it.

Reply:
Believe the entry should be explicitly removed from the map.

Reply:
The cancel func from the map has been implicitly removed or replaced when it is assigned in line 90 of `factory.go`:
```go
factory.tokenCancelFunc[tokenStr] = cancel
```
and the replaced one will be garbage collected eventually by Go.

Calling canceling function does not mean the cancel function has to be removed from the map even if go-routine is gone.  Whether the cancel function is effective or not is a different story or issue.   And to keep in the case of token-refreshing is still on-going, we want to use the latest cancelFunc, that is why the lines 82-90 of factory.go are doing that job. Hopefully this explanation is clean to you.

Reply:
Should only call a context's cancel function once.

You have at least once path through the code where you call cancel and don't update the map.  You should be removing the map entry in that specific case.

Reply:
I see where you are coming from now, so let me walk through a few cases and see if this makes sense:

1. token1 is new and the first time to call the client, so from line 82, the token1 is not in the map yet and that cancel() in line 83 won't be called. Then from line 87-90, if there is error from refreshToken() call, then cancel() gets called and the go-routine will be terminated. In this case, token1 is not even in the map.  So we are good in this case

2. token1 is already in the map, and the caller calls the NewSecretClient() again, so the line 83 cancel() gets call and will terminate the go-routine spawn in the previous call.  With that being said, if token1 is in the map, that mean it has successfully refreshToken in some previous run and thus line 90 had stored that token1 into map.  Now, in this second time, there are two cases:
(1) let say, if somehow, refreshToken is failing, then it will cancel the potential go-routine and in this case we want to remove the map entry  **<--- I guess this is where you were talking about.**  
(2) On the second case, if refreshToken is successfully done and thus a go-routine is spawn so that the logic in line 90 will take care of updating the map entry with the cancel func this time.

So I can add the logic to take care of that 2.(1) case just to make sure map state is good.  Thanks!